### PR TITLE
node scaling proposal

### PR DESCRIPTION
**Pull request by vangheem**
### Description
This PR write down a simple proposal to help us scale our node storage reads and writes.

It also implements a NIP(nucliadb improvement proposal) process.

### CODE REVIEW

File: .proposals/records/index-scaling-v1.md [hunk 1]
```diff
@@ -0,0 +1,235 @@
+# Index Scaling: Decoupling disk, reads/writes
```
Comment (jotare) [inline .proposals/records/index-scaling-v1.md:1]: I sounds nice! The only thing that worries me is the latency between write to gcs and read sync and search hitting a IndexReadReplica not available during this time.

When a write occurs on the system, all read replicas will be stale during: gcs write sync time + reader sync time. In this scenario, a search will potentially call a read replica and we'll need a second round trip to the index (at best) to obtain search results. The current API allow a `shards` parameter with a list of shard replicas to query and the frontend caches the searched shard. If the frontend caches a shard replica of a IndexReadReplica, two round trips are always ensured while not in sync. The number of potential hits to a stale read replica scales with the addition of new replicas.

However, within the assumption we'll have way more reads than writes, this shouldn't be too common.

File: .proposals/records/index-scaling-v1.md [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+*Goals*:
+- Allow short term scaling needs of IndexNode and search requests
+- Allow to scale reads separately from writes without rebalancing churn
+- Minimize impact of strategy on existing design to validate strategy without too much refactoring churn
+
+
+*Assumptions*:
+- Reads/writes should be scaled separately
+- (We will read much more than we will write). Is this true?
```
Comment (lferran) [inline .proposals/records/index-scaling-v1.md:19]: We should have metrics in place for this, but what I was assuming so far:

 - In general, a particular document will be written once and read (searched) many times
 - Another thing is the global write rates Vs search rates that the IndexNode may receive: They may be for sure different and it is a reason already to aim to scale both separately.

File: .proposals/README.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# NIPs(NucliaDB Improvements Proposals)
```
[the remark this answers is not in the document]
Comment (vangheem) [reply .proposals/README.md:1]: yeah, I can move it back to non `.`. I guess I just didn't like that it was tucked in with all our code modules.

[the remark this answers is not in the document]
Comment (vangheem) [reply .proposals/README.md:1]: oh, good point. I will update

**Comment by vangheem** [comment]
f2f meeting notes:

- concerns over time to search and would need to keep an eye on this metric and watch the stale reader replica retry implementation
- large shards under active writes will cause a lot of churn and writes. Can we keep up? Is it too much data? How big can they be? How much of a problem is it?
- look into event sourcing as another potential solution: store every indexing event in GCS and replaying to create another reader replica


Metrics to gather:
- time to search
- how often files are updated in an index and how large are they

**Pull request closed without merging**